### DISCUSSION TITLE
L1/L2 cache interaction

### DISCUSSION THREAD

**Discussion by gitting** [AzureAD/microsoft-identity-web#1879] [Q&A]
The scenario is multiple instances of `IConfidentialClientApplication` calling `AcquireTokenForClient` with the resulting token to be shared among all of them. The distributed cache is the answer and reading [this document](https://github.com/AzureAD/microsoft-identity-web/wiki/L1-Cache-in-Distributed-(L2)-Token-Cache) suggests the L1/L2 combination would be even better.

The expected flow.

Initialization:
- one instance starts up, gets a token from AAD, caches it in both memory and distributed
- second instance starts up, no in-memory token, checks the distributed cache, gets a token and if it's good then no calls to AAD
- more instances come up similarly

Ongoing:
1) Some instance discovers the token needs refreshing based on its L1
2) It checks the L2 to see if some other instance renewed the token
3) If not, then it obtains one from AAD and stores it in both caches
4) Other instances just use the newly updated L2 per 2.

However, from my testing it doesn't appear to work that way. In 2. above the instance does not check the distributed and goes straight to AAD and then updates both caches. All the other instances follow suit since they act strictly based on their in-memory caches. Hence, the only benefit of L2 cache is for any new instance initial coming up. The ongoing long-living instances have no benefit at all.

I can make L2 cache work as desired if I disable L1 cache via `MsalDistributedTokenCacheAdapterOptions.DisableL1Cache`. But that would obviously eliminate L1 benefits.

Is this the current and intended implementation or am I missing something? Thanks.

**Comment by jmprieur** [comment]
Are your instances of MSAL CCA in the same application? Or different?
Also do they have the same client ID?

**Reply by gitting** [reply]
It's a test exe with an `IDistributedCache` implementation against local filesystem. I just launch several instances of it so
- the app is constructed identically
- the process has its own memory cache but a common distributed one 

I also set it up with a custom `AppTokenProvider` so I have short token expirations. Each exe then loops asking for a token. And I observe the logging from my custom distributed implementation.

**Reply by gitting** [reply]
Taking a cursory look [this](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs#:~:text=protected%20override%20async%20Task%3Cbyte%5B%5D%3E%20ReadCacheBytesAsync(string%20cacheKey%2C%20CacheSerializerHints%20cacheSerializerHints)) appears to work as follows:

- if L1 is enabled and has a token it gets returned
- else L2 is consulted and L1 is set with what's found there

This would explain the behavior I've described. Initial L1 population benefits from L2, but subsequently the upstream (assuming MSAL) will make decisions based on the token it gets from its own dedicated memory cache.

The [Refresh](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs#:~:text=_distributedCache.RefreshAsync(cacheKey%2C%20cacheSerializerHints.CancellationToken)%2C) on the L2 with the L1 token (if one is found) doesn't look helpful. It could work if `Refresh` returned a token that would be refreshed, if needed, inside the L2 implementation. Then the L1 could be updated similar to [this](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs#:~:text=//%20back%20propagate%20to%20memory%20cache). But then the innards of MSAL doing the AAD interaction would have to be duplicated here - not a good idea.
